File: driver_model/driver_model.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Explorative code for studying loudspeaker physics.

This is not a software component to be used in larger engineering
designs as-is.

This algorithm will likely be integrated in
speaker_experiments/virtual_speakers in C++.
"""


import numpy
import scipy
import scipy.io
import scipy.io.wavfile
import scipy.ndimage


# Read the wave using scipy

samplerate, input_wav = scipy.io.wavfile.read('/tmp/1.wav')
input_wav = input_wav.astype(float)
#input_wav *= 1/32768 # 32 bit samples need scaling down

def fade(wav):
  for i in range(min(10000, len(wav))):
    mul = 1.0 - i / 10000.;
    wav[wav.shape[0] - i - 1][0] = 0
    wav[wav.shape[0] - i - 1][1] = 0

fade(input_wav)
print(input_wav.shape[0])

print("gaussian")
#blurred = scipy.ndimage.gaussian_filter1d(input_wav, 330.0, 0, 0)
#input_wav -= blurred
blurred = scipy.ndimage.gaussian_filter1d(input_wav, 440.0, 0, 0)
input_wav -= blurred
print("end gaussian")


output = numpy.zeros(input_wav.shape)

# parameters
# suspension strength is basically a correction level multiplier
suspension = 0.00034

# damping reduces the speed of the membrane passively as it
# emits energy or converts it to heat in the suspension deformations
damping = 0.99975

currpow = 0.0

hysteresis = 0
hysteresis_damping = 0.9
force_to_hysteresis = 9e-10
min_hys = 99
max_hys = -99

coil_heat = 0
coil_cooling = 0.9997
coil_heat_mul = 3e-15
max_heat = 0

displacement_mul = 1e-9

for c in range(input_wav.shape[1]):
  print("loopi", c)
  dpos = 0.0
  pos = 0.0
  for i in range(input_wav.shape[0]):
    force = 0
    if (i >= 2):
        force = (-0.01 * (input_wav[i-2][c] + input_wav[i][c]) +
               1.02 * input_wav[i-1][c])

    v = force
    for k in range(4):
      dpos *= damping
      dpos += 0.25 * force
      pos += 0.25 * dpos;
      v += 0.25 * suspension * pos
      pos *= 0.99825


    #print (v)
    coil_heat += abs(v * v)
    coil_heat *= coil_cooling

    hysteresis += force_to_hysteresis * v
    hysteresis *= hysteresis_damping

    min_hys = min(min_hys, hysteresis)
    max_hys = max(max_hys, hysteresis)
    #hysteresis = min(max(hysteresis, -0.005), 0.005)

    max_heat = max(max_heat, coil_heat)
    output[i][c] = v
    output[i][c] *= 1.0 - hysteresis
    output[i][c] *= 1.0 + coil_heat * coil_heat_mul
    #output[i][c] *= 1.0 + 1e-39 * abs(pos) ** 4


print("max stats", min_hys, max_hys, max_heat * coil_heat_mul)

output *= 0.3 / 32768.

scipy.io.wavfile.write('/tmp/o.wav', samplerate, output)

File: driver_model/driver_model.cc
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
This utility is intended to improve the driver dynamics of a wave
field synthesis loudspeaker prototype currently located on Jyrki's desk.

This utility works best with 48k sample rate.
*/

#include <stdio.h>
#include <stdlib.h>

#include <algorithm>
#include <sndfile.hh>
#include <string>
#include <vector>

struct Sound {
 public:
  std::vector<float> wav;
  int numsamples;
  int numchannels;
  int samplerate;
};

void Read(const std::string &path, Sound *snd) {
  SndfileHandle f(path);
  if (!f) {
    fprintf(stderr, "%s\n", f.strError());
    exit(-1);
  }
  int numchannels = f.channels();
  snd->numchannels = numchannels;
  snd->samplerate = f.samplerate();
  int numsamples = f.frames();
  snd->numsamples = numsamples;
  snd->wav.resize(numchannels * numsamples);

  printf("read %d channels, %d samples: size()=%zu\n", numchannels, numsamples,
         snd->wav.size());
  if (f.readf(snd->wav.data(), numsamples) != numsamples) {
    fprintf(stderr, "%s\n", f.strError());
    exit(-1);
  }
}

void Write(const std::string &path, const Sound &snd) {
  int numsamples = snd.numsamples;
  int numchannels = snd.numchannels;
  int samplerate = snd.samplerate;
  printf("write %d channels, %d samples\n", numchannels, numsamples);
  SndfileHandle f(path, /*mode=*/SFM_WRITE,
                  /*format=*/SF_FORMAT_WAV | SF_FORMAT_PCM_24,
                  /*channels=*/numchannels, /*samplerate=*/samplerate);
  f.writef(snd.wav.data(), snd.numsamples);
}


struct MultiChannelDriverModel {
  std::vector<float> pos;
  std::vector<float> dpos;
  void Initialize(size_t n) {
    pos.resize(n);
    dpos.resize(n);
  }
  void Convert(float *p, size_t n) {
    const float kSuspension = 0.00039;

    // damping reduces the speed of the membrane passively as it
    // emits energy or converts it to heat in the suspension deformations
    const float damping = 0.99999;

    const float kInputMul = 0.3;

    for (int k = 0; k < n; ++k) {
      float v = kInputMul * p[k];
      dpos[k] *= damping;
      dpos[k] += v;
      pos[k] += dpos[k];
      v += kSuspension * pos[k];
      pos[k] *= 0.99999; // Somewhat random position regularization.
      p[k] = v;
    }
  }
};


void DriverModel(int n, float *p, int stride) {
  MultiChannelDriverModel dm;
  dm.Initialize(1);

  for (int i = 0; i < n; i += stride) {
    dm.Convert(&p[i], 1);
  }
}

#include <stdlib.h>
#include <unistd.h>

bool SkipChannel(int c, const Sound &snd) {
  return snd.numchannels == 20 && (c == 0 || c == 1 || c == 10 || c == 11);
}

int main(int argc, char **argv) {
  if (argc != 3) {
    fprintf(stderr,
            "Usage: %s in.wav out.wav\n"
            "Only 32 bit float tracks [sox -e float -b 32] are supported\n",
            argv[0]);
    exit(1);
  }
  const char *wav_in = argv[1];
  const char *wav_out = argv[2];
  Sound snd;
  Read(wav_in, &snd);
  for (int c = 0; c < snd.numchannels; ++c) {
    if (SkipChannel(c, snd)) continue;
    fprintf(stderr, "processing channel %d\n", c);
    DriverModel(snd.wav.size(), &snd.wav[c], snd.numchannels);
  }
  Write(wav_out, snd);
  return 0;
}
